### ISSUE TITLE
Cannot read property 'type' of undefined.

### ISSUE THREAD

**Issue by ak151997galaxy** [expo/router#349]
### Summary

Hi!,
In app, we used 3 layouts - RootLayout, TabLayout, HomeLayout
and we used the useLocation Hook on the home screen and got this error and getRootState is undefined.
![Screenshot_2023-03-06-16-17-42-05_2e2efda142863eeb137780c265b51d87](https://user-images.githubusercontent.com/104891451/223090431-4a80847a-446a-417e-9a75-c26c28e8a6d2.jpg)

### Minimal reproducible example

for minimal reproducible example i created a repo.
https://github.com/ak151997galaxy/expo-router-tabs

**Comment by obrassard** [comment]
I have the same issue !

**Comment by EvanBacon** [comment]
Please update the repro to not use the private package `@galaxycard/react-native-turbo-utils`

**Comment by ak151997galaxy** [comment]
repo updated.

**Comment by boris-burgos** [comment]
I had the same problem and the reason was that my provider did not return the <Slot/> children. In case it helps.

**Comment by CodeWithAlexander** [comment]
adding a ? before .type in useToLinkPath seemed to fix the issue i don't know if it's the proper way to handle the issue but just thought i'd share

**Comment by Nikola-Milovic** [comment]
Since I am having a similar issue, I'll post my steps to recreate


### Stack trace 

```
Logs for your project will appear below. Press Ctrl+C to exit.
Android Bundling complete 2526ms
 ERROR  TypeError: Cannot read property 'type' of undefined

This error is located at:
    in App (created by Route(App))
    in Route (created by Route(App))
    in LocationProvider (created by Route(App))
    in Route(App) (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by QualifiedSlot)
    in QualifiedSlot (created by DefaultNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by QualifiedNavigator)
    in QualifiedNavigator (created by Navigator)
    in Navigator (created by DefaultNavigator)
    in DefaultNavigator (created by Route(DefaultNavigator))
    in Route (created by Route(DefaultNavigator))
    in LocationProvider (created by Route(DefaultNavigator))
    in Route(DefaultNavigator) (created by RootRoute)
    in RootRoute (created by ContextNavigator)
    in InitialRootStateProvider (created by ContextNavigator)
    in EnsureSingleNavigator
    in BaseNavigationContainer
    in ThemeProvider
    in NavigationContainerInner (created by NavigationContainer)
    in NavigationContainer (created by ContextNavigator)
    in RootRouteNodeProvider (created by ContextNavigator)
    in ContextNavigator (created by ExpoRoot)
    in RNCSafeAreaProvider (created by SafeAreaProvider)
    in SafeAreaProvider (created by ExpoRoot)
    in RNGestureHandlerRootView (created by GestureHandlerRootView)
    in GestureHandlerRootView (created by GestureHandler)
    in GestureHandler (created by ExpoRoot)
    in ExpoRoot (created by App)
    in App (created by withDevTools(App))
    in withDevTools(App) (created by ErrorOverlay)
    in ErrorToastContainer (created by ErrorOverlay)
    in ErrorOverlay
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent), js engine: hermes
 ERROR  TypeError: Cannot read property 'type' of undefined

This error is located at:
    in App (created by Route(App))
    in Route (created by Route(App))
    in LocationProvider (created by Route(App))
    in Route(App) (created by SceneView)
    in StaticContainer
    in EnsureSingleNavigator (created by SceneView)
    in SceneView (created by QualifiedSlot)
    in QualifiedSlot (created by DefaultNavigator)
    in PreventRemoveProvider (created by NavigationContent)
    in NavigationContent
    in Unknown (created by QualifiedNavigator)
    in QualifiedNavigator (created by Navigator)
    in Navigator (created by DefaultNavigator)
    in DefaultNavigator (created by Route(DefaultNavigator))
    in Route (created by Route(DefaultNavigator))
    in LocationProvider (created by Route(DefaultNavigator))
    in Route(DefaultNavigator) (created by RootRoute)
    in RootRoute (created by ContextNavigator)
    in InitialRootStateProvider (created by ContextNavigator)
    in EnsureSingleNavigator
    in BaseNavigationContainer
    in ThemeProvider
    in NavigationContainerInner (created by NavigationContainer)
    in NavigationContainer (created by ContextNavigator)
    in RootRouteNodeProvider (created by ContextNavigator)
    in ContextNavigator (created by ExpoRoot)
    in RNCSafeAreaProvider (created by SafeAreaProvider)
    in SafeAreaProvider (created by ExpoRoot)
    in RNGestureHandlerRootView (created by GestureHandlerRootView)
    in GestureHandlerRootView (created by GestureHandler)
    in GestureHandler (created by ExpoRoot)
    in ExpoRoot (created by App)
    in App (created by withDevTools(App))
    in withDevTools(App) (created by ErrorOverlay)
    in ErrorToastContainer (created by ErrorOverlay)
    in ErrorOverlay
    in RCTView (created by View)
    in View (created by AppContainer)
    in RCTView (created by View)
    in View (created by AppContainer)
    in AppContainer
    in main(RootComponent), js engine: hermes
```

### Steps to recreate

1. Create the demo project with expo router `npx create-react-native-app -t with-router`
2. Create `index.js` under `/app`

```javascript

import { Text, View } from 'react-native';
import { useRouter } from 'expo-router';

export default function App() {
    const router = useRouter()


    if (!HasSession()) {
        router.push('/auth')
    }

    return (
        <View
            style={{
                flex: 1,
                justifyContent: 'center',
                alignItems: 'center',
            }}>
            <Text onPress={
                () => {
                    router.push('/auth')
                }
            }>Main screen! (press to go to login)</Text>
        </View>
    );;
}


function HasSession() {
    return false
}
```

3. Create `auth.js` under `/app`
```javascript
import { Text, View } from 'react-native';

export default function App() {
    return (
        <View
            style={{
                flex: 1,
                justifyContent: 'center',
                alignItems: 'center',
            }}>
            <Text>Login screen!</Text>
        </View>
    );;
}
```

### Expected behaviour

I used a similar setup where I check the session of the user and then navigate to auth if the session is invalid. Albeit this was on web. 

### Current behaviour

Crashes with the above stack trace, to "fix" it, remove the `router.push` outside the `<Text onPress>` and then click the text, it will navigate as expected

**Comment by RyelBanfield** [comment]
I am having this issue as well. 

route.replace works fine if used in an onPress, however, it does not work when doing the below.

```typescript
const useProtectedRoute = (session: Session | null) => {
  const segments = useSegments();
  const router = useRouter();

  useEffect(() => {
    const inAuthGroup = segments[0] === "(auth)";

    if (!session && !inAuthGroup) router.replace("/");
    if (session && inAuthGroup) router.replace("/");
  }, [session, segments]);
};
``` 

This is the entire snippet.

```typescript
import { Text } from "@rneui/themed";
import { Session } from "@supabase/supabase-js";
import { useRouter, useSegments } from "expo-router";
import { createContext, useContext, useEffect, useState } from "react";
import { View } from "react-native";

import { supabase } from "../supabase";

const AuthContext = createContext<Session | null>(null);

export const useAuth = () => useContext(AuthContext);

const useProtectedRoute = (session: Session | null) => {
  const segments = useSegments();
  const router = useRouter();

  useEffect(() => {
    const inAuthGroup = segments[0] === "(auth)";

    if (!session && !inAuthGroup) router.replace("/");
    if (session && inAuthGroup) router.replace("/");
  }, [session, segments]);
};

export const Provider = ({ children }: { children: React.ReactNode }) => {
  const [session, setSession] = useState<Session | null>(null);
  const [loading, setLoading] = useState(true);

  useProtectedRoute(session);

  useEffect(() => {
    void supabase.auth.getSession().then(({ data: { session } }) => {
      setSession(session);
      setLoading(false);
    });

    supabase.auth.onAuthStateChange((_event, session) => {
      setSession(session);
      setLoading(false);
    });
  }, []);

  if (loading)
    return (
      <View>
        <Text>Loading</Text>
      </View>
    );

  return (
    <AuthContext.Provider value={session}>{children}</AuthContext.Provider>
  );
};
```

**Comment by ItsWendell** [comment]
After experimenting a little bit it seems to happen because the navigation state isn't fully initialized yet. Here's an workaround for using `router.push` in useEffects, here we verify whether we actually have a state key in the router, this seems to fix the problem for me. There might be better state properties to depend on than the key.

```ts

const useProtectedRoute = (session: Session | null) => {
  const segments = useSegments();
  const router = useRouter();
  const navigationState = useRootNavigationState();

  useEffect(() => {
    if (!navigationState?.key) {
      // Temporary fix for router not being ready.
      return;
    }
    const inAuthGroup = segments[0] === "(auth)";

    if (!session && !inAuthGroup) router.replace("/");
    if (session && inAuthGroup) router.replace("/");
  }, [session, segments, navigationState?.key]);
};

```

**Comment by RyelBanfield** [comment]
Thank you @ItsWendell. This at works for me :)

**Comment by Nikola-Milovic** [comment]
@RyelBanfield @ItsWendell  https://expo.github.io/router/docs/guides/auth/ Did you see this guide?

**Comment by RyelBanfield** [comment]
@Nikola-Milovic yes, if you read my code from earlier, you would see it's based on the example from that guide.

**Comment by Nikola-Milovic** [comment]
Yeah but it works without adding anything, at least for me. @RyelBanfield

**Comment by RyelBanfield** [comment]
That's interesting. I think I do remember it working for me at first, then I'm not sure when but it started giving me the error that lead to me looking for this issue. @Nikola-Milovic

**Comment by israelidanny** [comment]
Same exact thing happened to me, and then navigationState thing did the trick.

**Comment by mbyio** [comment]
I encountered this when returning `<Redirect>`'s in the root layout. I was not able to find a workaround, because `useRootNavigationState` always returns null.

**Comment by RobertUpchurch** [comment]
If anyone finds it helpful, I found that i encountered this issue even with the suggestion above when I did not render {children} in all routes. This I believe is because the children is where it is rendering a <Stack /> or <Slot /> object. Without {children} the navigationRoot was always undefined. The fix was to ensure that a <Stack /> or <Slot /> is rendered in all routes.

I then decided to keep the <SplashScreen /> up until after i determined the authentication status or had already routed to the auth page.

```typescript
return (
    <AuthContext.Provider value={{ ... }}>
      {(isAuthPage || !!user) && <SplashScreen />}
      {children}
    </AuthContext.Provider>
  );
```

**Comment by Yuukis7** [comment]
I have similar issue but in my case `TypeError: Cannot read property 'segments' of undefined` caused by only calling `useSegments()` makes that workaround `if (!navigationState?.key) return;` isn't possible and error occurs only on build application (using Expo Go works fine, so to check if change helped I need 10 minutes each time to build app firstly). Sorry for not opening different issue and making it more detailed, but it's my first time using issue's tab and don't know that github's stuff. If anyone had same issue and knows solution or what might cause that I'd grateful and sorry again.

**Comment by julian-gargicevich** [comment]
Has anyone found a proper way to fix it?

```
if (!navigationState?.key) {
      // Temporary fix for router not being ready.
      return;
    }
```
Doesn't always work since navigationState can stay undefined

**Comment by zRelux** [comment]
Yup same issue

**Comment by trtin** [comment]
Same issue

**Comment by willenleal** [comment]
Having the same issue, here is another minimal reproducible issue repo: https://github.com/willenleal/expo-router-authprovider-issue

**Comment by manuelxaguilar** [comment]
Having the same issue, also following https://expo.github.io/router/docs/guides/auth/

**Comment by marklawlor** [comment]
Should be fixed in v2, but possibly blocked by https://github.com/expo/router/pull/794

**Comment by MahathirMohammadShuvo** [comment]
> That's interesting. I think I do remember it working for me at first, then I'm not sure when but it started giving me the error that lead to me looking for this issue. @Nikola-Milovic

/

> That's interesting. I think I do remember it working for me at first, then I'm not sure when but it started giving me the error that lead to me looking for this issue. @Nikola-Milovic

exact same happen with me. at first it was working fine. but suddenly it start giving this error.

**Comment by marklawlor** [comment]
I'm going to close this issue, as the original reported issue has been resolved. If your still experiencing this issue, can you please create a new issue in https://github.com/expo/expo
